`General`: Add external LLM Usage user settings

florian-glombik opened the pull request:
### Checklist
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).


#### Server
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.


#### Client
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) stubs
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.


### Motivation and Context
Once accepting the Iris user settings, there was no way to decline them again. This will now be possible with a newly added section `Iris` in the user settings.

In a follow-up PR, we will have a look at which data we can clear and delete, if a user who once used Iris decides to then revoke the acceptance of the conditions. This is not the scope of this PR.

![image](https://github.com/user-attachments/assets/a1df3b42-806f-45b6-90c2-2fb45ebe3739)


### Description
- Adds an Iris user section
- Declaring constants at the top of the file (and make them readonly)
- Do not display decline button on course dashboard, as it has no effect there

### Steps for Testing
1. Log in to Artemis
2. Navigate to the dashboard
3. Consent to the usage of iris (accept)
4. Navigate to the iris user settings
5. See that you have an option to revoke your consent
6. Navigate back to the dashboard
7. See that you can consent to the usage of iris again
8. (Verify there is no decline button on the dashboard, as it does not have an effect there)

### Testserver States
You can manage test servers using [Helios](https://helios.aet.cit.tum.de/). Check environment statuses in the [environment list](https://helios.aet.cit.tum.de/repo/69562331/environment/list). To deploy to a test server, go to the [CI/CD](https://helios.aet.cit.tum.de/repo/69562331/ci-cd) page, find your PR or branch, and trigger the deployment.

### Review Progress
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

### Test Coverage


### Screenshots
#### english
![image](https://github.com/user-attachments/assets/cdee0a08-d762-4e7d-8f7c-3c8c7b1005d5)

![image](https://github.com/user-attachments/assets/5365699b-f227-452a-8ab6-336066d5b044)

#### german
![image](https://github.com/user-attachments/assets/db774573-3a19-43a6-9476-3fbd56597c9a)

![image](https://github.com/user-attachments/assets/bb85bc81-9e4e-4784-8a75-f4e78740a6b7)

#### dashboard only has the accept button now (decline button did not do anything)
![image](https://github.com/user-attachments/assets/0bc4884c-db2b-434d-ac86-51f106b5e87e)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Added a user settings page for managing consent to external large language model (LLM) usage, allowing users to accept or decline at any time.
  - Introduced clear status messages and navigation for external LLM usage settings in both English and German.
- **Enhancements**
  - Updated consent dialogs and messages to use improved wording for clarity.
  - Consent status is now accurately reflected and can be updated throughout the application.
  - Refined consent handling logic to allow repeated acceptance or declination without errors.
  - Updated multiple components and services to use the new consent update method with explicit acceptance state.
  - Added conditional display of decline button in chatbot components for better user control.
- **Bug Fixes**
  - Fixed issues where consent status was not properly recorded or displayed after user action.
- **Tests**
  - Added and updated unit and integration tests to cover new consent scenarios and ensure correct behavior.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

maximiliansoelch reviewed (dismissed):
I am questioning the naming of the added setting section as this is not only concerning Iris, but also other systems in the edutelligence space relying on LLMs like for example Athena.

I suggest to name this new setting more generic, e.g. External LLM Usage.

florian-glombik commented:
> I am questioning the naming of the added setting section as this is not only concerning Iris, but also other systems in the edutelligence space relying on LLMs like for example Athena.
> 
> I suggest to name this new setting more generic, e.g. External LLM Usage.

Sure, I adjusted it accordingly - I removed the check for iris to be enabled in this case aswell. Do you think that we should always keep it in the user settings? Otherwise which other features in Artemis require consent regarding external LLM usage?

jfr2102 reviewed (approved):
code looks good, I left inline comments

tobias-lippert commented on src/main/java/de/tum/cit/aet/artemis/core/web/UserResource.java:
```diff
@@ -125,19 +127,24 @@ public ResponseEntity<UserInitializationDTO> initializeUser() {
+
+        boolean isExternalLLMUsageAlreadyAccepted = user.getExternalLLMUsageAcceptedTimestamp() != null;
+        if (isExternalLLMUsageAlreadyAccepted && acceptExternalLLMUsageDTO.accepted()) {
             return ResponseEntity.badRequest().build();
```
and i think the idea of put mappings is to be idempotent. Otherwise, make it a PostMapping.

ahmetsenturk reviewed (dismissed):
tested on TS3, also by requesting AI feedback (which also requires external LLM confirmation), works fine 👍

tobias-lippert commented on src/main/webapp/app/core/course/overview/exercise-details/request-feedback-button/request-feedback-button.component.ts:
```diff
@@ -31,7 +31,21 @@ import { getAllResultsOfAllSubmissions } from 'app/exercise/shared/entities/subm
+    private readonly modalService = inject(NgbModal);
+    private readonly participationWebsocketService = inject(ParticipationWebsocketService);
+
+    protected readonly faPenSquare = faPenSquare;
```
might be just me, but I'd switch the order of protected and private fields.

florian-glombik replied:
Let's discuss this next monday 😄

tobias-lippert commented on src/test/java/de/tum/cit/aet/artemis/core/user/AccountResourceIntegrationTest.java:
```diff
@@ -497,14 +498,32 @@ void acceptExternalLLMUsageSuccessful() throws Exception {
+        AcceptExternalLLMUsageDTO acceptExternalLLMUsageDTO = new AcceptExternalLLMUsageDTO(false);
+        request.put("/api/core/users/accept-external-llm-usage", acceptExternalLLMUsageDTO, HttpStatus.OK);
+
+        // Verify no timestamp was set
```
llm-generated comment?

tobias-lippert commented on src/test/java/de/tum/cit/aet/artemis/core/user/AccountResourceIntegrationTest.java:
```diff
@@ -497,14 +498,32 @@ void acceptExternalLLMUsageSuccessful() throws Exception {
+    @Test
+    @WithMockUser(username = AUTHENTICATEDUSER)
+    void declineExternalLLMUsageSuccessful() throws Exception {
+        // Create user in repo with null timestamp
```
omit comment?

tobias-lippert reviewed (approved):
just some nits. Nothing that blocks the PR.

tobias-lippert reviewed (changes requested):
related client test fails

tobias-lippert reviewed (approved):
code

florian-glombik commented on src/main/webapp/i18n/de/exercise-actions.json:
```diff
@@ -71,7 +71,7 @@
             "externalLLMUsage": {
                 "popUpLabel": "Bestätigung der externen LLM-Verwendung",
-                "popUpMessage": "Indem du fortfährst, bestätigst du, dass du mind. 18 Jahre alt bist und dass du damit einverstanden bist, dass deine Interaktionen mit Artemis und die notwendigen kontextbezogenen Daten, z.B. dein Code, von Microsoft und OpenAI (d.h. außerhalb unseres Universitätsrechenzentrums) verarbeitet werden.",
+                "popUpMessage": "Indem du akzeptierst, bestätigst du, dass du mind. 18 Jahre alt bist und dass du damit einverstanden bist, dass deine Interaktionen mit Artemis und die notwendigen kontextbezogenen Daten, z.B. dein Code, von Microsoft und OpenAI (d.h. außerhalb unseres Universitätsrechenzentrums) verarbeitet werden.",
```
I am not sure about legal requirements here 😅
